redis7/redis.go

```diff
@@ -0,0 +1,371 @@
+}
+
+func (t tx) Store() stoabs.KVStore {
+	return t.store
+}
+
+func (t tx) Unwrap() interface{} {
+	return t.writer
+}
+
+type shelf struct {
+	name   string
+	prefix string
+	reader redis.Cmdable
+	// writer holds the Redis command writer for writing to the database.
+	// It is separate from the reader because writes use a transactional Redis pipeline, which buffers the commands.
+	// Buffering read commands would result in unexpected behaviour where the reads are only executed on transaction commit.
+	writer redis.Cmdable
+	store  *store
+}
+
+func (s shelf) Put(key stoabs.Key, value []byte) error {
+	if err := s.store.checkOpen(); err != nil {
+		return err
+	}
+	return s.writer.Set(context.TODO(), s.toRedisKey(key), value, 0).Err()
+}
+
+func (s shelf) Delete(key stoabs.Key) error {
+	return s.writer.Del(context.TODO(), s.toRedisKey(key)).Err()
+}
+
+func (s shelf) Get(key stoabs.Key) ([]byte, error) {
+	result, err := s.reader.Get(context.TODO(), s.toRedisKey(key)).Result()
+	if err == redis.Nil {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	return []byte(result), nil
+}
+
+func (s shelf) Iterate(callback stoabs.CallerFn, keyType stoabs.Key) error {
+	var cursor uint64
+	var err error
+	var keys []string
+	for {
+		scanCmd := s.reader.Scan(context.TODO(), cursor, s.toRedisKey(stoabs.BytesKey(""))+"*", int64(resultCount))
+		keys, cursor, err = scanCmd.Result()
+		if err != nil {
+			return err
+		}
+		if len(keys) == 0 {
+			// Nothing to iterate over
+			return nil
+		}
+		_, err := s.visitKeys(keys, callback, keyType, false)
+		if err != nil {
+			return err
+		}
+		if cursor == 0 {
+			// Done
+			break
+		}
+	}
+	return nil
+}
+
+func (s shelf) Range(from stoabs.Key, to stoabs.Key, callback stoabs.CallerFn, stopAtNil bool) error {
+	keys := make([]string, 0, resultCount)
+	// Iterate from..to (start inclusive, end exclusive)
+	var numKeys = 0
+	for curr := from; !curr.Equals(to); curr = curr.Next() {
+		if curr.Equals(to) {
+			// Reached end (exclusive)
+			break
+		}
+		keys = append(keys, s.toRedisKey(curr))
+		// We don't want to perform requests that are really large, so we limit it at page size
+		if numKeys >= resultCount {
+			proceed, err := s.visitKeys(keys, callback, from, stopAtNil)
+			if err != nil || !proceed {
+				return err
+			}
+			keys = make([]string, 0, resultCount)
+			numKeys = 0
+		} else {
+			numKeys++
+		}
+	}
+	_, err := s.visitKeys(keys, callback, from, stopAtNil)
+	return err
+}
+
+func (s shelf) visitKeys(keys []string, callback stoabs.CallerFn, keyType stoabs.Key, stopAtNil bool) (bool, error) {
+	values, err := s.reader.MGet(context.TODO(), keys...).Result()
+	if err != nil {
+		return false, err
+	}
+	for i, value := range values {
+		if values[i] == nil {
+			// Value does not exist (anymore), or not a string
+			if stopAtNil {
+				return false, nil
+			}
+			continue
+		}
+		key, err := s.fromRedisKey(keys[i], keyType)
+		if err != nil {
+			return true, err
```

Comment:
Why return `true` and error?

Reply:
Doesn't actually matter because `err` is checked beforehand, but it's not consistent with the other cases. I'll make if `false` and add a godoc for the function (because it's a bit oddball).